### PR TITLE
Update Darwin framework API to get connected device

### PR DESCRIPTION
**Pull request by pan-apple**
#### Problem
Darwin framework's `getConnectedDevice` is not idiomatic. The error should not be provided as an out parameter, and should be returned via completion handler. Also, the completion handler should be the last argument.

#### Change overview
Updated the API and it's usage.

#### Testing
Ran iOS CHIPTool app to test the updated API.

### CODE REVIEW

**Comment by mspang** [comment]
Can you clarify something for me - What is the lifetime of the result of device or a "connected device" supposed to be? Who's responsible for destroying it?

**Comment by pan-apple** [comment]
> Can you clarify something for me - What is the lifetime of a device or a "connected device" supposed to be? Who's responsible for destroying it?

The old API `getPairedDevice` had left the lifetime maintenance up to the user/app of the API. i.e. the caller was supposed to know when to release the device.

With `getConnectedDevice()`, I am hoping to contain the lifetime to `OnDeviceConnected` callback. It's not fully achieved yet, as there are code blocks which are still using the old API, and some code is storing the device object pointer provided in the callback.

Any suggestions?

**Comment by mspang** [comment]
> > Can you clarify something for me - What is the lifetime of a device or a "connected device" supposed to be? Who's responsible for destroying it?
> 
> The old API `getPairedDevice` had left the lifetime maintenance up to the user/app of the API. i.e. the caller was supposed to know when to release the device.
> 

But this worked, right?

device = GetDevice(nodeId)
device = GetDevice(nodeId)
ReleaseDevice(device)

If caller owns, that should not be allowed, because it leads to the following sequence when composing two otherwise reasonable pieces of code together:

device1 = GetDevice(nodeId)
device2 = GetDevice(nodeId)
ReleaseDevice(device1)
device2->Oops(); // Oops, use-after-free

> With `getConnectedDevice()`, I am hoping to contain the lifetime to `OnDeviceConnected` callback. It's not fully achieved yet, as there are code blocks which are still using the old API, and some code is storing the device object pointer provided in the callback.
> 
> Any suggestions?

That doesn't seem right to me. What if I want to send a sequence of commands to a device?

There needs to be a large design discussion about these APIs..

@msandstedt @mrjerryjohns

**Comment by pan-apple** [comment]
> > > Can you clarify something for me - What is the lifetime of a device or a "connected device" supposed to be? Who's responsible for destroying it?
> > 
> > 
> > The old API `getPairedDevice` had left the lifetime maintenance up to the user/app of the API. i.e. the caller was supposed to know when to release the device.
> 
> But this worked, right?
> 
> device = GetDevice(nodeId)
> device = GetDevice(nodeId)
> ReleaseDevice(device)
> 
> If caller owns, that should not be allowed, because it leads to the following sequence when composing two otherwise reasonable pieces of code together:
> 
> device1 = GetDevice(nodeId)
> device2 = GetDevice(nodeId)
> ReleaseDevice(device1)
> device2->Oops(); // Oops, use-after-free
> 
> > With `getConnectedDevice()`, I am hoping to contain the lifetime to `OnDeviceConnected` callback. It's not fully achieved yet, as there are code blocks which are still using the old API, and some code is storing the device object pointer provided in the callback.
> > Any suggestions?
> 
> That doesn't seem right to me. What if I want to send a sequence of commands to a device?
> 
> There needs to be a larger design discussion about these APIs..
> 
> @msandstedt @mrjerryjohns

Yes, I agree. We need wider discussion to decide the API behavior.

**Comment by msandstedt** [comment]
> > > Can you clarify something for me - What is the lifetime of a device or a "connected device" supposed to be? Who's responsible for destroying it?
> > 
> > 
> > The old API `getPairedDevice` had left the lifetime maintenance up to the user/app of the API. i.e. the caller was supposed to know when to release the device.
> 
> But this worked, right?
> 
> device = GetDevice(nodeId)
> device = GetDevice(nodeId)
> ReleaseDevice(device)
> 
> If caller owns, that should not be allowed, because it leads to the following sequence when composing two otherwise reasonable pieces of code together:
> 
> device1 = GetDevice(nodeId)
> device2 = GetDevice(nodeId)
> ReleaseDevice(device1)
> device2->Oops(); // Oops, use-after-free
> 
> > With `getConnectedDevice()`, I am hoping to contain the lifetime to `OnDeviceConnected` callback. It's not fully achieved yet, as there are code blocks which are still using the old API, and some code is storing the device object pointer provided in the callback.
> > Any suggestions?
> 
> That doesn't seem right to me. What if I want to send a sequence of commands to a device?
> 
> There needs to be a larger design discussion about these APIs..
> 
> @msandstedt @mrjerryjohns

I have found the fact the controller maintains an allocation pool for devices to be problematic in the first place:

* node-ID keying only works in single-fabric scenarios
* a fixed-size pool is undesirable for unconstrained platforms
* awkward / impossible to usefully extend `chip::controller::Device` because the controller allocates them
* more...

In fact, I don't really understand why we need to have a fixed, central Device object pool at all.  For instance, device list persistence appears to mostly be an artifact of maintaining PASE session counters across boots.  Without this, what is the remaining need for this type of persistence?  To cache mDNS-resolved addresses?  That could easily be a separate module.  Fundamentally, the actual way to persist device-to-device associations is by distributing certificates and writing ACLs.  `chip::controller::Device` object serialization and persistence doesn't do any of that for you.

What I would really like to be able to do is construct Device objects as I need to and control lifecycle directly.

It does appear that `chip::controller::Device` already provides some of what is needed to allow this.  It still gets several things from `chip::controller::DeviceController` though:
 
* Transport Manager bound to a port
* AdminPairingTable / AdminPairingInfo / OperationalCredentialSet
* Message Counter Manager
* Session Manager resource pool
* Exchange Manager resource pool
* Session ID allocator

But I don't see how any of these controller-provided facilities lead to the requirement that the `chip::controller::DeviceController` allocate Device objects or manage their lifecycles.  This is greatly complicating things.

This also introduces unnecessary constraints.  It should be legal for a node to instantiate up to 4096 concurrent sessions with a single peer.  But we've decided through the controller implementation and its fixed Device (peer node) pool with no more than a single entry per peer that only a single session per peer should be supported.

TLDR: `chip::controller::Device` would be more accurately described as a peer session.  These do not need to be persisted by the core sdk and it is very awkward that they are currently allocated in a fixed nodeid-keyed pool inside the `chip::controller::DeviceController` object.  Apps should be able to instantiate these objects at will and control lifecycle directly.

@kianooshkarami, @samadDotDev

**Comment by mspang** [comment]
> > > > Can you clarify something for me - What is the lifetime of a device or a "connected device" supposed to be? Who's responsible for destroying it?
> > > 
> > > 
> > > The old API `getPairedDevice` had left the lifetime maintenance up to the user/app of the API. i.e. the caller was supposed to know when to release the device.
> > 
> > 
> > But this worked, right?
> > device = GetDevice(nodeId)
> > device = GetDevice(nodeId)
> > ReleaseDevice(device)
> > If caller owns, that should not be allowed, because it leads to the following sequence when composing two otherwise reasonable pieces of code together:
> > device1 = GetDevice(nodeId)
> > device2 = GetDevice(nodeId)
> > ReleaseDevice(device1)
> > device2->Oops(); // Oops, use-after-free
> > > With `getConnectedDevice()`, I am hoping to contain the lifetime to `OnDeviceConnected` callback. It's not fully achieved yet, as there are code blocks which are still using the old API, and some code is storing the device object pointer provided in the callback.
> > > Any suggestions?
> > 
> > 
> > That doesn't seem right to me. What if I want to send a sequence of commands to a device?
> > There needs to be a larger design discussion about these APIs..
> > @msandstedt @mrjerryjohns
> 
> I have found the fact the controller maintains an allocation pool for devices to be problematic in the first place:
> 
> * node-ID keying only works in single-fabric scenarios
> * a fixed-size pool is undesirable for unconstrained platforms
> * awkward / impossible to usefully extend `chip::controller::Device` because the controller allocates them
> * more...
> 
> In fact, I don't really understand why we need to have a fixed, central Device object pool at all. For instance, device list persistence appears to mostly be an artifact of maintaining PASE session counters across boots. Without this, what is the remaining need for this type of persistence? To cache mDNS-resolved addresses? That could easily be a separate module. Fundamentally, the actual way to persist device-to-device associations is by distributing certificates and writing ACLs. `chip::controller::Device` object serialization and persistence doesn't do any of that for you.
> 
> What I would really like to be able to do is construct Device objects as I need to and control lifecycle directly.
> 
> It does appear that `chip::controller::Device` already provides some of what is needed to allow this. It still gets several things from `chip::controller::DeviceController` though:
> 
> * Transport Manager bound to a port
> * AdminPairingTable / AdminPairingInfo / OperationalCredentialSet
> * Message Counter Manager
> * Session Manager resource pool
> * Exchange Manager resource pool
> * Session ID allocator
> 
> But I don't see how any of these controller-provided facilities lead to the requirement that the `chip::controller::DeviceController` allocate Device objects or manage their lifecycles. This is greatly complicating things.
> 
> This also introduces unnecessary constraints. It should be legal for a node to instantiate up to 4096 concurrent sessions with a single peer. But we've decided through the controller implementation and its fixed Device (peer node) pool with no more than a single entry per peer that only a single session per peer should be supported.
> 
> TLDR: `chip::controller::Device` would be more accurately described as a peer session. These do not need to be persisted by the core sdk and it is very awkward that they are currently allocated in a fixed nodeid-keyed pool inside the `chip::controller::DeviceController` object. Apps should be able to instantiate these objects at will and control lifecycle directly.
> 
> @kianooshkarami, @samadDotDev

Agree on these points and we've come to similar conclusions.. in fact we did some brief brainstorming what kind of API we'd like for session establishment & cluster commands over here last week.

On the persistence model that, what exists currently is indeed limiting. I don't think a model where the SDK itself is issuing writes of key -> value pairs that are opaque to the application is a good fit for us generally speaking. Too many policy concerns come into play and just like with our "external" synchronization discussion, we should move toward "external" persistence management.

**Review by jmartinez-silabs — APPROVED** [review]

**Review by anush-apple — APPROVED** [review]

File: src/darwin/Framework/CHIP/CHIPDeviceController.mm [hunk 1]
```diff
@@ -374,21 +374,24 @@ - (BOOL)isDevicePaired:(uint64_t)deviceID error:(NSError * __autoreleasing *)err
...
 
     dispatch_async(_chipWorkQueue, ^{
         CHIPDeviceConnectionBridge * connectionBridge = new CHIPDeviceConnectionBridge(completionHandler, queue);
-        errorCode = connectionBridge->connect(self->_cppCommissioner, deviceID);
+        CHIP_ERROR errorCode = connectionBridge->connect(self->_cppCommissioner, deviceID);
 
-        if ([self checkForError:errorCode logMsg:kErrorGetPairedDevice error:error]) {
+        NSError * error;
+        if ([self checkForError:errorCode logMsg:kErrorGetPairedDevice error:&error]) {
             // Errors are propagated to the caller thru completionHandler.
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPDeviceController.mm:395]:
```suggestion
            // Errors are propagated to the caller through completionHandler.
```

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
@saurabhst @msandstedt @mspang @andy31415 @LuDuda could you please take a look with the understanding that we do need to work out better APIs here but in the meantime attempts to use a Device without waiting for it to set up CASE lead to random breakage and this PR is attempting to address that by doing that waiting appropriately.

**Review by mspang — APPROVED** [review]

**Comment by msandstedt** [comment]
We spent some time playing with the controller and device objects and have had some success.  I want to share this to hopefully provide some actionable advice.

The current architecture that encloses the pool of device objects within the controller object is understandable from an example application perspective, but unworkable for actual ecosystem implementations.  Problems with this:
- 64-connected-device pool size: insufficient for many applications
- fixed-size device pool: many applications will require dynamic allocation
- controller-internal device allocation: many applications will need to extend the device object
- 128-paired-device pool: node/node associations are created by commissioning nodes and writing ACLs; beyond this, ecosystems will have their own needs and approaches when tracking commissioned devices; it's impossible for the sdk to be written in such a way that these needs are addressed; thus tracking and persisting paired device in a core component of the sdk does not make sense
- node-ID keying: precludes any type of multi-fabric support for a controller

To address these and other problems, in at least one codebase, we are moving toward a model of device allocation and lifecycle management that is external to the controller object. We have not found any other way to work around the controller object's limitations.

To achieve this, we're using the controller object to bootstrap application state, and thereafter constructing device objects dynamically and initializing them with the output of `chip::ControllerDeviceController::GetControllerDeviceInitParams`.  We've found that with only minor modifications to the device object, this works well.

As to what the plan would be for the sdk's controller object, I think we need to come to a point where there is a clear separation between spec-informed behavior, and other code that is only relevant for example controllers and commissioners.  Here is an incomplete list of the things the controller / commissioner objects are doing that likely will not make sense for ecosystem-specific implementations:

- paired device persistence
- controller OpCred bootstrapping (self commissioning)
- allocation of device session objects
- commissioning statefulness

What I would really like to see is a separation of the controller / commissioner objects into stateless helpers and the stateful wrappers that the examples can call into.  In leui of this, ecosystem implementations will likely need to implement their own controller and commissioner objects.  That is effectively what we are doing with the approach I described above.

The reason I call this out here is that this PR is attempting to address a problem that ought to be very specific to the example controllers.  In at least one of our applications, the controller's management of device objects isn't something we're able to accommodate in the first place.

Maybe this is good news.  Maybe the more general spec-conformant object is chip::Controller::Device and that's what ecosystems should leverage in their implementations (as I described above).  My concern though is that this forces more development out of the sdk than should be necessary.  The commissioner is more complex than the controller.  It feels like we ought to be able to provide some sort of general-purpose helper object for this.  And we can anticipate that Darwin and Android will be frequent targets for the commissioning role.  It feels unfortunate to be in a place where all implementations on these targets must roll their own commissioner object.
